porch/pkg/engine/engine.go

```diff
@@ -546,3 +554,76 @@ func (m *mutationReplaceResources) Apply(ctx context.Context, resources reposito
 
 	return repository.PackageResources{Contents: new}, task, nil
 }
+
+func healConfig(old, new map[string]string) (map[string]string, error) {
+	// The processor simply replaces the old resource list with new one,
+	// matching resources by names and copying their IDs over
+	ff := runtimeutil.FunctionFilter{
+		Run: func(reader io.Reader, writer io.Writer) error {
+			items, err := (&kio.ByteReader{
+				Reader:             reader,
+				WrappingAPIVersion: kio.ResourceListAPIVersion,
+				WrappingKind:       kio.ResourceListKind,
+			}).Read()
+			if err != nil {
+				return fmt.Errorf("failed to parse old config with resource IDs: %w", err)
+			}
+
+			var copyID kio.FilterFunc = func(r []*yaml.RNode) ([]*yaml.RNode, error) {
+				for _, n := range r {
+					for _, original := range items {
+						if n.GetNamespace() == original.GetNamespace() &&
+							n.GetName() == original.GetName() &&
+							n.GetApiVersion() == original.GetApiVersion() &&
+							n.GetKind() == original.GetKind() {
+							id, err := original.Pipe(yaml.GetAnnotation(kioutil.IdAnnotation))
```

Comment:
@justinsb what algorithm did you use to match resources in the manual implementation of comment healing? I tried to reuse kpt machinery here and this seemed ok-ish to start with.

alternative would be to assign IDs when reading package resources though kpt libraries may not play nicely with it.